blog/_posts/2021-10-19-scala-3.1.0-released.md

```diff
@@ -0,0 +1,113 @@
+---
+layout: blog-detail
+post-type: blog
+by: Paweł Marks, VirtusLab
+title: Scala 3.1.0 released!
+---
+Hello from the Scala 3 team! It has already been six weeks since we have announced the release candidate for the first minor version after the initial release of Scala 3. Now, after two more RCs, we can confidently promote Scala 3.1.0-RC3 to a new stable release of the language. In other words, Scala 3.1.0 is officially out!
+
+You don't need to worry about having to migrate your projects to the new version. Scala 3 has strong compatibility guarantees, and all code working in 3.0.2 will also work in 3.1. On the other hand, switching to Scala 3.1 allows you to use many improvements and a handful of newly stabilized APIs.
+
+## What's new in 3.1
+
+### New experimental feature: safer exceptions
+
+A new experimental feature that allows declaring and checking which exceptions can be thrown. It relies on the effects as implicit capabilities pattern.
+
+You can opt-in using following import.
+
+```scala
+import language.experimental.saferExceptions
+```
+
+Now it is possible to mark types with exception that cna be thrown during the evalutaion:
+
+```scala
+def f(x: Double): Double canThrow LimitExceeded =
+  if x < limit then f(x) else throw LimitExceeded())
+```
+
+You can read more in [the document proposing this feature](https://github.com/lampepfl/dotty/blob/release-3.1.0/docs/docs/reference/experimental/canthrow.md).
+
+### Efficient bytecode for matching over strings
+
+When Scala 3.0 has been compiling pattern matches with `Int` as a scrutinee type, it was emitting bytecode containing very efficient instructions: `lookupswitch` or `tableswitch`. Now in Scala 3.1, similar behavior was implemented for matching over `String` literals.
+
+```scala
+fruit match
+  case "apple"  => 1
+  case "orange" => 2
+  case "banana" => 3
+  case _        => 0
+```
+
+will now generate `lookupswitch` over hash codes of `String` literals instead of a chain of conditional expressions. This behavior is consistent with what Java is doing and more performant than the old approach.
+
+### Support for `-Wconf` and `@nowarn`
+
+We have added a `-Wconf` compiler flag that allows filtering and configuring compiler warnings (silence them, or turn them into errors).
+
+We have also integrated [the silencer plugin](https://github.com/ghik/silencer) into the compiler, which allows suppressing warnings locally using the `@nowarn` annotation.
+
+`-Wconf` and `@nowarn` work largely the same way as in Scala 2, but some filters for selecting warnings are different. For the more details see the output of `-Wconf:help` flag.
```

Comment:
That output is rather lengthy. To do a detailed comparison of the Scala 2 output with the Scala 3 output would be difficult — unless you're unusually eagle-eyed, you'd have to pipe them through `diff`. It might be better to just actually tell the reader what the differences are?